src/System.Runtime.Extensions/src/System/IO/StreamWriter.cs

```diff
@@ -253,12 +251,22 @@ private void Flush(bool flushStream, bool flushEncoder)
                 }
             }
 
-            int count = _encoder.GetBytes(_charBuffer, 0, _charPos, _byteBuffer, 0, flushEncoder);
-            _charPos = 0;
-            if (count > 0)
+            int count = 0;
+            byte[] byteBuffer = ArrayPool<byte>.Shared.Rent(_encoding.GetMaxByteCount(_charPos));
```

Comment:
Wouldn't it be better to rent a more stable array size? (as to improve reuse of shared arrays)
Why not always rent an array of size `_encoding.GetMaxByteCount(_charLen)`?
It'll consume more memory, but likelyhood of the same pooled array being reused should be much greater.

Reply:
Would cause more contention on a set bucket size's lock; as the defaults would generally be used so it would always be renting from the 4096 bucket.

Also when the 4096 bucket runs out, it would move up to the next so then be allocating 8192 arrays, followed by 16384 arrays, so increase memory consumption.

Are you thinking for CPU cache?

Reply:
I understand, but then again that's what you do for the char buffer?

Reply:
The char buffer is longer lived as it buffers/coalesces smaller `Write`s into a single Stream.Write.

However, the char buffer does get returned on a `Stream` flush and then rented again on next `Write` so it isn't held for the whole lifespan of the `StreamWriter`. Though it is less frequently rented and returned, which would hopefully reduce contention.

Also when you create a `StreamWriter` with buffer size of N currently, you may expect it to have and array of `N` chars 
`2 * N bytes`
but may be surprised that it actually allocates
 `2 * N + 4 * N bytes = 6 * N bytes` 
when you factor in the `byte[]` for UTF-8 encoding for example.

So a seemingly innocuous statement like
```csharp
using (var sw = new StreamWriter(httpContext.Response.Body, _UTF8encoding))
{
    sw.Write("Hello, World!");
}
```
Will currently allocate 6 x 1024 = 6144 (6kB) of arrays to write "Hello World!" to the stream; or if it was given a file path 6 x 4096 = 24576 (24kB) to a file.

Reply:
Sure, but when compared with the implementation before this PR using a longer lived buffer for byte would not increase allocations, would hardly increase contention (at least not more than when renting the char buffer), but would surely cause less GC pressure.

However, I trust your judgement on this (more than mine) :-)

Reply:
Also is lifetimes of the arrays; the threadpool has 8 arrays per core at any particular size; so if there are 1024 requests all being processed on an 8 core machine (64 per bucket); if they sit on them the pool will new up 960 extra arrays (although some will come from larger buckets); then when the requests are done; it will discard 960 of the 1024 arrays.

More granular use will allow more overlapped use, and hopefully not need any extra than the threadpool would normally have.